service/history/workflow/mutable_state_impl.go

```diff
@@ -765,6 +770,14 @@ func (ms *MutableStateImpl) GetNamespaceEntry() *namespace.Namespace {
 	return ms.namespaceEntry
 }
 
+// AddHistoryEvent adds any history event to this workflow execution.
+// The provided setAttributes function should be used to set the attributes on the event.
+func (ms *MutableStateImpl) AddHistoryEvent(t enumspb.EventType, setAttributes func(*history.HistoryEvent)) *history.HistoryEvent {
+	event := ms.hBuilder.AddHistoryEvent(t, setAttributes)
+	ms.writeEventToCache(event)
```

Comment:
For later: I think we should allow plugins/components customizing this.

Reply:
I'd like to know why that's important, I don't have the context. But we can discuss when there's a use case.